java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,933 @@
+import java
+private import RegexFlowConfigs
+
+// In all ranges handled by this library, `start` is inclusive and `end` is exclusive.
+/**
+ * A string literal that is used as a regular expression.
+ */
+abstract class RegexString extends StringLiteral {
+  /** Holds if a character set starts between `start` and `end`, including any negation character (`^`). */
+  private predicate charSetStart0(int start, int end) {
+    this.nonEscapedCharAt(start) = "[" and
+    (if this.getChar(start + 1) = "^" then end = start + 2 else end = start + 1)
+  }
+
+  /** Holds if the character at `pos` marks the end of a character class. */
+  private predicate charSetEnd0(int pos) {
+    this.nonEscapedCharAt(pos) = "]" and
+    /* special case: `[]]` and `[^]]` are valid char classes.  */
+    not this.charSetStart0(_, pos)
+  }
+
+  /**
+   * Gets the nesting depth of character classes at position `pos`
+   */
+  private int charSetDepth(int pos) {
+    exists(this.getChar(pos)) and
+    result =
+      max(int j |
+        j = 0 or
+        j =
+          count(int i | i < pos and this.charSetStart0(i, _)) -
+            count(int i | i < pos and this.charSetEnd0(i))
+      )
+  }
+
+  /** Hold if a top-level character set starts between `start` and `end`. */
+  predicate charSetStart(int start, int end) {
+    this.charSetStart0(start, end) and
+    this.charSetDepth(start) = 0
+  }
+
+  /** Holds if a top-level character set ends at `pos`. */
+  predicate charSetEnd(int pos) {
+    this.charSetEnd0(pos) and
+    this.charSetDepth(pos) = 1
+  }
+
+  /**
+   * Holds if there is a top-level character class beginning at `start` (inclusive) and ending at `end` (exclusive)
+   *
+   * For now, nested character classes are approximated by only considering the top-level class for parsing.
+   * This leads to very similar results for ReDoS queries.
+   */
+  predicate charSet(int start, int end) {
+    exists(int inner_start, int inner_end |
+      this.charSetStart(start, inner_start) and
+      not this.charSetStart(_, start)
+    |
+      end = inner_end + 1 and
+      inner_end > inner_start and
+      this.charSetEnd(inner_end) and
+      not exists(int mid | this.charSetEnd(mid) | mid > inner_start and mid < inner_end)
```

Comment:
Simplify: replace with `inner_end = min(...)`.

Reply:
The `min` version will likely be more efficient, but still quadratic in the number of top-level char classes. We can make it linear by keeping track of the depth=0 opening bracket in the rank-index traversal we need to do to calculate depth, and then immediately match up the top-level opening and closing brackets.